pkg/skaffold/build/gcb/status.go

```diff
@@ -0,0 +1,272 @@
+// It reduces the number of requests made to the GCB API by using a single `List` call for
+// all concurrently running builds for a project instead of multiple per-job `Get` calls.
+// The statuses are polled with an independent exponential backoff strategy per project until success, cancellation or failure.
+// The backoff duration for each `projectID` status query is reset with every new build request.
+type statusManager interface {
+	getStatus(ctx context.Context, projectID string, buildID string) <-chan result
+}
+
+// statusManagerImpl implements the `statusManager` interface
+type statusManagerImpl struct {
+	results     map[jobID]chan result
+	resultMutex sync.RWMutex
+	requests    chan jobRequest
+}
+
+// jobID represents a single build job
+type jobID struct {
+	projectID string
+	buildID   string
+}
+
+// jobRequest encapsulates a single build job and it's context.
+type jobRequest struct {
+	jobID
+	ctx context.Context
+}
+
+// result represents a single build job status
+type result struct {
+	jobID
+	status *cloudbuild.Build
+	err    error
+}
+
+func getStatusManager() statusManager {
+	reporterOnce.Do(func() {
+		reporter = &statusManagerImpl{
+			results:  make(map[jobID]chan result),
+			requests: make(chan jobRequest, 1),
+		}
+		// start processing all job requests in new goroutine.
+		go reporter.run()
+	})
+	return reporter
+}
+
+// poller sends the `projectID` on channel `C` with an exponentially increasing period for each project.
+type poller struct {
+	timers   map[string]*time.Timer   // timers keep the next `Timer` keyed on the projectID
+	backoffs map[string]*wait.Backoff // backoffs keeps the current `Backoff` keyed on the projectID
+	reset    chan string              // reset channel receives any new `projectID` and resets the timer and backoff for that projectID
+	remove   chan string              // remove channel receives a `projectID` that has no more running jobs and deletes its timer and backoff
+	trans    chan string              // trans channel pushes to channel `C` and sets up the next trigger
+	C        chan string              // C channel triggers with the `projectID`
+}
+
+func newPoller() *poller {
+	p := &poller{
+		timers:   make(map[string]*time.Timer),
+		backoffs: make(map[string]*wait.Backoff),
+		reset:    make(chan string),
+		remove:   make(chan string),
+		trans:    make(chan string),
+		C:        make(chan string),
+	}
+
+	go func() {
+		for {
+			select {
+			case projectID := <-p.remove:
+				if b, found := p.timers[projectID]; found {
+					b.Stop()
+				}
+				delete(p.timers, projectID)
+				delete(p.backoffs, projectID)
+			case projectID := <-p.reset:
+				if b, found := p.timers[projectID]; found {
+					b.Stop()
+				}
+				p.backoffs[projectID] = NewStatusBackoff()
+				p.timers[projectID] = time.AfterFunc(p.backoffs[projectID].Step(), func() {
+					p.trans <- projectID
+				})
+			case projectID := <-p.trans:
+				go func() {
+					p.C <- projectID
```

Comment:
if we already blocked on the `projectID := <-p.trans` call in the `case` statement, does this need to be in a `go func()`?